Overhaul keyword handling

nnethercote opened the pull request:
The compiler's list of keywords has some problems.
- It contains several items that aren't keywords.
- The order isn't quite right in a couple of places.
- Some of the names of predicates relating to keywords are confusing.
- rustdoc and rustfmt have their own (incorrect) versions of the keyword list.
- `AllKeywords` is unnecessarily complex.

r? @jieyouxu

nnethercote committed:
Move `gen` in the keyword list.

`gen` is an edition-specific keyword used in unstable Rust, and so
belongs with `try` (as `is_unused_keyword_conditional` indicates).

Also, the cases in `is_unused_keyword_conditional` should be in
alphabetical order, to match the keyword list.

These changes don't affect the behaviour of any of the `Symbol::is_*`
functions.

nnethercote committed:
Simplify `AllKeywords`.

It's a verbose reinvention of a range type, and can be cut down a lot.

nnethercote committed:
Only have one source of truth for keywords.

`rustc_symbol` is the source of truth for keywords.

rustdoc has its own implicit definition of keywords, via the
`is_doc_keyword`. It (presumably) intends to include all keywords, but
it omits `yeet`.

rustfmt has its own explicit list of Rust keywords. It also (presumably)
intends to include all keywords, but it omits `await`, `builtin`, `gen`,
`macro_rules`, `raw`, `reuse`, `safe`, and `yeet`. Also, it does linear
searches through this list, which is inefficient.

This commit fixes all of the above problems by introducing a new
predicate `is_any_keyword` in rustc and using it in rustdoc and rustfmt.
It documents that it's not the right predicate in most cases.

bors commented:
:umbrella: The latest upstream changes (presumably #134269) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
Best reviewed one commit at a time.

jieyouxu commented:
@nnethercote I haven't looked at this in detail just yet, does this PR adjust how contextual keywords like `safe` are handled?

nnethercote commented:
> @nnethercote I haven't looked at this in detail just yet, just a quick question: does this PR adjust how _contextual_ keywords like `safe` are handled?

No. The only potential behavioural changes are that some of the (non)keywords (e.g. `kw::DollarCrate`, `kw::StaticLifetime`) that were accepted by the `is_*` predicates no longer are. But that had no effect on any tests. None of the contextual keywords were accepted by the predicates. More details in the individual commit messages.

nnethercote committed:
Improve comments on `Keywords`.

In particular, clarify which predicates apply to which keywords.

nnethercote commented:
@petrochenkov: your definition of "keyword" doesn't align with that given by the language Reference, or the Ferrocene spec, or the terminology used by the compiler's error messages. But I have zero interest in litigating that so I have removed the things you objected to. I kept `used_keywords` as is because I think it's just as good there as somewhere else. I kept the name `is_any_keyword` but I added comments explaining that you shouldn't use it.

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #134414) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
>your definition of "keyword" doesn't align with that given by the language Reference, or the Ferrocene spec, or the terminology used by the compiler's error message

If there's a mismatch then we should ideally work with the reference people to harmonize the spec and the reality.
- Keywords being a subset of identifiers (*) is something that must be in the spec, it's a behavior that is publicly visible to users through proc macro APIs. Same applies to `_` being a keyword.
  - (*) Identifiers that are either always or conditionally reserved by the language proper (not by individual macros/attributes).
- Lifetime identifiers and keywords are an implementation peculiarity that shouldn't leak into the public spec (e.g. proc macros do not have them), we should ideally do the refactoring mentioned in https://github.com/rust-lang/rust/pull/134253#discussion_r1887284680 and get rid of them.
- The `Symbol::is_special` group is purely an internal detail, we can do whatever is more convenient (e.g. keep things as is).

petrochenkov commented on compiler/rustc_span/src/symbol.rs:
```diff
@@ -20,18 +20,25 @@ mod tests;
-    // If you modify this list, adjust `is_special`, `is_used_keyword`/`is_unused_keyword`
-    // and `AllKeywords`.
+    // This list includes things that are definitely keywords (e.g. `if`),
+    // things that are definitely not keywords (e.g. the empty symbol,
```
```suggestion
    // a few things that are definitely not keywords (e.g. the empty symbol,
```

petrochenkov commented on compiler/rustc_span/src/symbol.rs:
```diff
@@ -20,18 +20,25 @@ mod tests;
     Keywords {
-        // Special reserved identifiers used internally for elided lifetimes,
-        // unnamed method parameters, crate root module, error recovery etc.
+        // "Special" things.
```
Could you keep the removed part of the comment?

>"Special" reserved symbols are symbols that are not identifiers lexically, but are commonly used internally in AST/HIR `Ident`ifiers for elided lifetimes, ... ... ...

petrochenkov commented on compiler/rustc_span/src/symbol.rs:
```diff
@@ -83,23 +91,29 @@ symbols! {
         Try:                "try", // >= 2018 Edition only
 
-        // Special lifetime names
+        // "Special" lifetime names.
```
```suggestion
        // "Lifetime keywords" (regular keywords with prepended `'`) used in `Ident`ifiers in lifetimes.
```

petrochenkov commented:
r=me after rebase and some comment tweaks.
@rustbot author

nnethercote commented:
Rebased and comments addressed.

@bors r=petrochenkov

bors commented:
:pushpin: Commit 9753bf4025d7d6ef5805fa0edb66845f26891fe6 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 9753bf4025d7d6ef5805fa0edb66845f26891fe6 -->
<!-- homu: {"type":"Approved","sha":"9753bf4025d7d6ef5805fa0edb66845f26891fe6","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #134243) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@bors r=petrochenkov

bors commented:
:pushpin: Commit 6de550cc68d5eb7dc5013987c87859e2f589ac12 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 6de550cc68d5eb7dc5013987c87859e2f589ac12 -->
<!-- homu: {"type":"Approved","sha":"6de550cc68d5eb7dc5013987c87859e2f589ac12","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
